From aa0f79a965bf54351ff2d86f1d7a8c36d7f59c62 Mon Sep 17 00:00:00 2001
From: Natanael Copa <ncopa@alpinelinux.org>
Date: Thu, 20 Oct 2016 11:37:47 +0200
Subject: [PATCH] Support musl libc without libbsd

libbsd is not very portable and does not work with musl libc on arm.
When on linux we only require libbsd if GNU libc is detected.
---
 GNUmakefile |  4 ++++
 config.h    | 18 ++++++++++++++++--
 extern.h    |  2 ++
 3 files changed, 22 insertions(+), 2 deletions(-)

diff --git a/GNUmakefile b/GNUmakefile
index c6a605d..af2b815 100644
--- a/GNUmakefile
+++ b/GNUmakefile
@@ -25,7 +25,11 @@ OBJS 	 = acctproc.o \
 
 ifeq ($(shell uname), Linux)
 # Compiling on Linux.
+ifeq ($(shell $(CC) -dM -E - < /dev/null | grep __GLIBC__),__GLIBC__)
+# require libbsd for GNU libc, but not for musl libc
+CFLAGS  += -DHAVE_LIBBSD
 LIBBSD	 = -lbsd
+endif
 CFLAGS	+= -I/usr/local/include/libressl
 LDFLAGS += -L/usr/local/lib
 OBJS	+= util-portable.o \
diff --git a/config.h b/config.h
index 3aa6563..423382f 100644
--- a/config.h
+++ b/config.h
@@ -17,8 +17,21 @@
  */
 #ifdef __linux__
 # define _GNU_SOURCE
-# include <bsd/stdlib.h>
-# include <bsd/string.h>
+# ifdef HAVE_LIBBSD
+#  include <bsd/stdlib.h>
+#  include <bsd/string.h>
+# else
+#  ifndef __BEGIN_DECLS
+#   define __BEGIN_DECLS
+#  endif
+#  ifndef __END_DECLS
+#   define __END_DECLS
+#  endif
+#  include <errno.h>
+   static inline const char *getprogname() {
+	return program_invocation_short_name;
+   }
+# endif
 # include <grp.h>
 #endif
 
@@ -41,3 +54,4 @@ int	setresuid(gid_t, gid_t, gid_t);
 #include <netinet/in.h>
 #include <resolv.h>
 #endif
+
diff --git a/extern.h b/extern.h
index ada0063..ebf38aa 100644
--- a/extern.h
+++ b/extern.h
@@ -17,6 +17,8 @@
 #ifndef EXTERN_H
 #define EXTERN_H
 
+#include <unistd.h> /* for pid_t */
+
 #ifndef PATH_VAR_EMPTY
 #define	PATH_VAR_EMPTY "/var/empty"
 #endif
-- 
2.10.1

